WIP: Client: confirm before dumping binary data to terminal

DMRobertson opened the pull request:
For your consideration: a patch which implements the QOL feature described in #59524.

<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Improvement


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
The CLI client now warns the user if binary output will be printed to the terminal, and requests confirmation before continuing.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

I had a quick grep through the docs, and I couldn't see any examples using `--output-format Parquet` directly.

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

Closes #59524.

----

Notes on the first-time contributor experience:

- CMake wanted to use `make` instead of `ninja` on my System (Fedora 38); I had to pass `-DC_MAKE_MAKE_PROGRAM=ninja` to workaround this.
  - Final invocation was `cmake -G Ninja -DENABLE_LIBRARIES=0 -DCMAKE_MAKE_PROGRAM=ninja -DCMAKE_BUILD_TYPE=Debug ..`.
- My laptop is apparently underpowered for compiling things locally. Even with `-D_ENABLELIBRARIES=0` I found it took a while to get things going. (In particular the linker took ~6 minutes, though subsequent linker runs took ~1 minute.)
- Wasn't sure if there were standard linters to run: I tried `clang-tidy` and didn't get any huge warning signs. Maybe CI will find more to report, though.

Notes on the implementation:

- I tried to add this check before we communicate with the server or load any history etc, to provide immediate feedback.
- I was originally going to use the `LineReader` instance created inside `ClientBase::runInteractive`, assuming that this was "the done thing". But this wasn't a good fit: that reader class manages history, multiline inputs, other delimiters etc, none of which we care about for a simple `y/n` prompt.
- I tested manually by running `programs/clickhouse-local --output-format Parquet` in the build directory.
  - Inputs like `y`, `Y`, `  y  ` are accepted and bring me to the prompt;
  - inputs like `n`, `N`, `  n     ` end the program;
  - an empty input (immediate newline) or whitespace only issues a reprompt.
  - other inputs like `a` or `sadoasdnasodu` also issue a reprompt.
  - inputting EOF with Ctrl-D or emitting a SIGINT with Ctrl-C ends the program.
- I wondered if some kind of option/env var to disable the confirmation prompt (e.g. `--allow-tty-binary-output`) but didn't want to open up that can of worms.

DMRobertson committed:
Client: confirm before dumping Parquet to terminal

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=59740) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ClickHouse/ClickHouse?pullRequest=59740) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ClickHouse/ClickHouse?pullRequest=59740) it.</sub>

DMRobertson committed:
Replace accidental tabs with spaces

That's what I get for editing with one IDE and then switching to vim at
the end.

alexey-milovidov commented:
Ok, but it should not treat Parquet as special. There are many other binary formats, such as Avro, ORC, RowBinary, Native, MsgPack, Protobuf, BSON, etc...

DMRobertson commented:
Thanks. I did check the description of `--output-format` in `--help` to see what the available output formats were, but found no guidance. I guess https://clickhouse.com/docs/en/interfaces/formats is the authoritative list of formats here?

Does the source already have a notion of which formats are binary? If not, where is the right place to add this---a bool flag inside `FormatFactory::Creators` perhaps?

DMRobertson committed:
Have the FormatFactory track which output formats are binary

DMRobertson commented:
> If not, where is the right place to add this---a bool flag inside `FormatFactory::Creators` perhaps?

I've pushed another commit assuming this is the case. I looked through all the calls to `registerOutputFormat.*` to see which emit binary data and marked those formats as such (though please check my working).

I sanity-checked this locally:

```
$ programs/clickhouse-local --output-format Parquet
ClickHouse local version 24.2.1.1.

This will dump binary output to your terminal. Did you mean to redirect to a file with e.g. '>output'?
Dump binary output to terminal? (y/n): ^C
$ programs/clickhouse-local --output-format Native
ClickHouse local version 24.2.1.1.

This will dump binary output to your terminal. Did you mean to redirect to a file with e.g. '>output'?
Dump binary output to terminal? (y/n): n
Code: 236. DB::Exception: Binary output to terminal declined by user. (ABORTED)

$ programs/clickhouse-local --output-format CSV
ClickHouse local version 24.2.1.1.

hostname :) 
```

robot-ch-test-poll commented:
<!-- automatic status comment for PR #59740 from DMRobertson/ClickHouse:dmr/confirm-binary-data-dump -->
*This is an automated comment for commit 9d276de0e44ce7454d0ee98d5b2b84279f598495 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/59740/9d276de0e44ce7454d0ee98d5b2b84279f598495/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/59740/9d276de0e44ce7454d0ee98d5b2b84279f598495/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

alexey-milovidov committed:
Update RawBLOBRowOutputFormat.cpp

alexey-milovidov commented on src/Client/ClientBase.cpp:
```diff
@@ -2372,6 +2373,32 @@ void ClientBase::runInteractive()
     if (print_time_to_stderr)
         throw Exception(ErrorCodes::BAD_ARGUMENTS, "time option could be specified only in non-interactive mode");
 
+    if (stdout_is_a_tty && FormatFactory::instance().checkIfOutputFormatIsBinary(format))
```
1. Missing tests.
2. This check will happen at the client startup and only in the interactive mode rather than for a query.

Check what happens if a user issued multiple queries, e.g. `SELECT 1 FORMAT TSV`, `SELECT 1 FORMAT Native`.
